guides/v2.1/howdoi/patch_from_github.md

```diff
@@ -0,0 +1,62 @@
+---
+layout: default
+group: howdoi
+title: Create a GitHub patch for a Composer installation
+version: 2.2
+github_link: howdoi/patch_from_github.md
+functional_areas:
+  - Install
+  - System
+  - Setup
+---
+
+Sometimes it takes a while for us to include a bug fix made on GitHub in a Magento 2 Composer release. In the meantime, you can create a patch from GitHub and use the `composer-patches` plugin to apply it to your Composer-based Magento 2 installation.
+
+<div class="bs-callout bs-callout-warning" markdown="1">
+Be careful applying an unreleased patch. Do comprehensive testing before deploying it.
+</div>
+
+## Create a patch
+
+To create a patch file from a GitHub commit or pull request, append `.patch` to the url, [https://github.com/magento/magento2/commit/2d31571f1bacd11aa2ec795180abf682e0e9aede.patch](https://github.com/magento/magento2/commit/2d31571f1bacd11aa2ec795180abf682e0e9aede.patch).
```

Comment:
This URL shows you the patch file, but doesn't actually create it for your purposes. I *think* you copy the contents, and paste it into the file described in line 31. 

When  take the link, I see this at the top:

```
From 2d31571f1bacd11aa2ec795180abf682e0e9aede Mon Sep 17 00:00:00 2001
From: Ievgen Sentiabov <isentiabov@magento.com>
Date: Wed, 31 Aug 2016 18:46:44 +0300
Subject: [PATCH] MAGETWO-56934: Checkout page freezes when ordering with
 Authorize.net with invalid credit card

 - Added fail() method call
---
 app/code/Magento/Payment/view/frontend/web/js/view/payment/iframe.js | 1 +
 1 file changed, 1 insertion(+)
```

Does any of this need to be deleted? It's certainly not essential, but maybe the `composer-patches` tool knows to ignore it.

Reply:
@dmanners, deferring to you on this

Reply:
Alternatively, you can also add `.diff` to the url which results in a smaller file then the `.patch` variant.
But both work absolutely fine. People should pick the one they like best.

And yes, it's best if you then download that file and put it in the directory mentioned below: `patches/composer` (this directory is very arbitrary btw, you can choose which one you want to use yourself, not sure if this needs to be mentioned?)

Some other additions here:

- Sometimes a PR doesn't contain the full fix, because sometimes a member of the magento community engineering team adds some extra code to the PR. If you would then add `.patch` or `.diff` to a PR, you won't get the full fix.
An example of this is this PR: https://github.com/magento/magento2/pull/15265, the correct fix is the commit mentioned at the very bottom (https://github.com/magento/magento2/commit/d9dc66cfc5e1988bd8673c2cfdd1de2b85f0a3c6) instead of the changes of the PR.

So maybe we'll need to add some section about how you find the correct commit (or even multiple commits) for a certain fix. Which will probably be pretty complex to write down...

- If a fix is old, it's hard to find a single merge commit with only a fix for that problem, this is due to how the magento developers used to merge branches, this has improved more in recent months, and you'll usually find a single commit you can use for a patch.
But if you can't find a single commit and you have multiple commits, what I usually do then is:
  * clone the Magento 2 repo
  * checkout the version I want to create a patch for
  * cherry-pick all commits you want to apply (and solve merge conflicts if they arise)
  * do a `git diff 2.1.4...HEAD > my-patch-file.diff` (where `2.1.4` is the version used in the second step)

- It also will regularly happen that a commit won't apply cleanly to a particular version of Magento, happens most because there was some other code changed in another commit which is missing in that particular version of Magento somewhere around the same lines of that patch.
If this happens, I also follow the 4 steps outlined above to create a new patch, but you will certainly have to solve merge conflicts in that case.

Feel free to ignore these things above if this would make the documentation more complex then it needs to. It might scare users or confuse them ...